src/kokkos_abstraction.hpp

```diff
@@ -214,470 +286,561 @@ inline void kokkos_dispatch(ParallelScanDispatch, Args &&...args) {
-                    function, std::forward<Args>(args)...);
+    return std::is_same_v<V, IndexRange> ? 2 : 1;
   }
-}
 
-template <typename Tag, typename Pattern, typename Function, class... Args>
-inline typename std::enable_if<sizeof...(Args) <= 1, void>::type
-par_dispatch(Pattern p, const std::string &name, DevExecSpace exec_space,
-             const IndexRange &r, const Function &function, Args &&...args) {
-  par_dispatch<Tag>(p, name, exec_space, r.s, r.e, function, std::forward<Args>(args)...);
-}
+  template <size_t N>
+  using Rank_t = std::integral_constant<size_t, N>;
 
-// 2D loop using MDRange loops
-template <typename Tag, typename Function, class... Args>
-inline typename std::enable_if<sizeof...(Args) <= 1, void>::type
-par_dispatch(LoopPatternMDRange, const std::string &name, DevExecSpace exec_space,
-             const int jl, const int ju, const int il, const int iu,
-             const Function &function, Args &&...args) {
-  Tag tag;
-  kokkos_dispatch(tag, name,
-                  Kokkos::Experimental::require(
-                      Kokkos::MDRangePolicy<Kokkos::Rank<2>>(
-                          exec_space, {jl, il}, {ju + 1, iu + 1}, {1, iu + 1 - il}),
-                      Kokkos::Experimental::WorkItemProperty::HintLightWeight),
-                  function, std::forward<Args>(args)...);
-}
+  using bound_variants = std::variant<IndexRange, IndexRange &>;
+  using bound = base_type<T>;
+  using LaunchBounds = GetLaunchBounds<TypeList<Args...>>;
 
-template <typename, typename>
-class FlatFunctor;
+ public:
+  using value =
+      typename std::conditional<is_BoundType<bound>(),
+                                insert_type_list_t<T, typename LaunchBounds::value, 0>,
+                                TypeList<>>::type;
+  using NumInds =
+      std::conditional_t<is_BoundType<bound>(),
+                         Rank_t<NumBnds<bound>() + LaunchBounds::NumInds::value>,
+                         Rank_t<NumBnds<bound>()>>;
+};
 
-template <typename F, typename... Args>
-auto MakeFlatFunctor(F &function, Args... args) {
-  return FlatFunctor<F, decltype(&F::operator())>(function, std::forward<Args>(args)...);
-}
+template <typename>
+struct DispatchSignature {};
+
+template <typename Index, typename... AllArgs>
+struct DispatchSignature<TypeList<Index, AllArgs...>> {
+ private:
+  using LB = GetLaunchBounds<TypeList<Index, AllArgs...>>;
+  using TL = TypeList<Index, AllArgs...>;
 
-template <typename Function, typename R, typename T, typename Index, typename... FArgs>
-class FlatFunctor<Function, R (T::*)(Index, Index, Index, FArgs...) const> {
-  int NjNi, Ni, kl, jl, il;
+ public:
+  using LaunchBounds = typename LB::value;
+  using Rank = std::integral_constant<size_t, LB::NumInds::value / 2>;
+  using Function = typename TL::template type<LaunchBounds::n_types>;
+  using Args = typename TL::template continuous_sublist<LaunchBounds::n_types + 1,
+                                                        TL::n_types - 1>;
```

Comment:
```
/Users/runner/work/parthenon/parthenon/src/utils/type_list.hpp:69:39: error: 'ContinuousSublist' is a private member of 'parthenon::TypeList<int, unsigned long, (lambda at /Users/runner/work/parthenon/parthenon/src/prolong_restrict/pr_loops.hpp:69:69)>'
  using continuous_sublist = decltype(ContinuousSublist<Start, End>());
                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/runner/work/parthenon/parthenon/src/kokkos_abstraction.hpp:367:38: note: in instantiation of template type alias 'continuous_sublist' requested here
  using Args = typename TL::template continuous_sublist<LaunchBounds::n_types + 1,
```
@lroberts36 it seems the macOS CI compiler takes issue with the `continuous_sublist` typedef `decltype`ing a private member function. Is there an obvious workaround besides making `ContinuousSublist` public?

Reply:
Hm, I don't know an obvious way to fix this aside from what you suggest (i.e. making `ContinuousSublist` public). I think that would be fine, but maybe we should rename `ContinuousSublist` to `ContinuousSublistImpl`.